servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java

```diff
@@ -71,74 +74,59 @@ public static ConcurrentSubscription wrap(Subscription subscription) {
-                // instead we just set to MIN_VALUE again and try to re-acquire the lock in case we raced again.
-                if (n < 0) {
-                    subscriptionRequestQueueUpdater.set(this, Long.MIN_VALUE);
-                } else {
-                    subscriptionRequestQueueUpdater.accumulateAndGet(this, n,
-                            FlowControlUtils::addWithOverflowProtectionIfNotNegative);
-                }
-                if (!subscriptionLockOwnerUpdater.compareAndSet(this, null, currentThread)) {
-                    return;
-                }
-                // We previously added our n contribution to the queue, but now that we have acquired the lock
-                // we are responsible for draining the queue.
-                n = subscriptionRequestQueueUpdater.getAndSet(this, 0);
-                if (n == 0) {
-                    // It is possible that the previous consumer has released the lock, and drained the queue before we
-                    // acquired the lock and drained the queue. This means we have acquired the lock, but the queue has
-                    // already been drained to 0. We should release the lock, try to drain the queue again, and then
-                    // loop to acquire the lock if there are elements to drain.
-                    subscriptionLockOwner = null;
-                    n = subscriptionRequestQueueUpdater.getAndSet(this, 0);
-                    if (n == 0) {
-                        return;
-                    } else {
-                        continue;
-                    }
+            pendingDemand = mapInvalidRequestN(n);
+        } else {
+            for (;;) {
+                final long prevPendingDemand = pendingDemand;
+                if (prevPendingDemand < 0 ||
+                        pendingDemandUpdater.compareAndSet(this, prevPendingDemand,
+                                prevPendingDemand + min(Long.MAX_VALUE - prevPendingDemand, n))) {
+                    break;
                 }
             }
-            if (n < 0) {
-                subscription.cancel();
-                return; // Don't set subscriptionLockOwner = 0 ... we don't want to request any more!
+        }
+        Throwable delayedCause = null;
+        for (;;) {
+            final long acquireId = acquirePendingReentrantLock(subscriptionLockUpdater, this);
+            if (acquireId == 0) {
+                break;
             }
+
             try {
-                subscription.request(n);
+                final long prevPendingDemand = pendingDemandUpdater.getAndSet(this, 0);
+                if (prevPendingDemand == CANCELLED) {
+                    subscription.cancel();
+                } else if (prevPendingDemand != 0) {
+                    subscription.request(prevPendingDemand);
+                }
+            } catch (Throwable cause) {
+                if (delayedCause == null) {
+                    delayedCause = cause;
+                }
```

Comment:
should we use `addSuppressed(...)` if `delayCause != null` ?

Reply:
lets discuss on https://github.com/apple/servicetalk/pull/1011#discussion_r408440046